### PR TITLE
add Gmsh mesh file parser

### PR DESCRIPTION
**Pull request by j8xixo12**
In this commit added mesh file io for modmesh, it lets modmesh has capabilty to read third-party mesh generator generated file and convert the content to a data structure that allow modmesh can use it to do some calculation.

Currently mesh io only support Gmsh format.

### CODE REVIEW

File: modmesh/app/sample_mesh.py [hunk 1]
```diff
@@ -83,6 +83,25 @@ def help_other(set_command=False):
         view.mgr.pycon.command = cmd.strip()
 
 
+def help_mesh_viewer(path, set_command=False):
```
Comment (j8xixo12) [inline modmesh/app/sample_mesh.py:86]: I also added a sample command in sample_mesh app for, once mesh viewer app is completed, then this sample command can be removed.

**Comment by yungyuc** [comment]
This PR contains almost 1000 lines of diff and will not be easy to review.  Would it be possible to break it into smaller pieces?  For review efficiency a batch of review should be around 300 lines of diff.

File: modmesh/core.py [hunk 2]
```diff
@@ -64,6 +64,7 @@
     'METAL_BUILT',
     'metal_running',
     'HAS_VIEW',
+    'Gmsh',
```
Comment (yungyuc) [inline modmesh/core.py]: This is a class and should be put before constants, ordered in the same convention as other classes or functions.

File: cpp/modmesh/io/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (yungyuc) [inline cpp/modmesh/io/CMakeLists.txt]: `io` the name conflict with Python `io` module.  It is unnecessary ambiguity.  Please rename this directory as `inout`.

File: cpp/modmesh/io/gmsh.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+#include <unordered_map>
+#include <functional>
+
+#include <modmesh/base.hpp>
+#include <modmesh/mesh/mesh.hpp>
+#include <modmesh/buffer/buffer.hpp>
+
+namespace modmesh
+{
+namespace IO
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: namespace should not use uppercase, and it should not have the unnecessary ambiguity to the Python module named `io`. 

In addition, `IO` is a pretty short name and may conflict with symbols in foreign code base.

Please rename it to `inout`.

File: cpp/modmesh/io/gmsh.hpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+{
+namespace Gmsh
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: Please do not use uppercase in namespace. Beside, I don't think it is necessary to have 3 levels of namespace. Gmsh is just one of the format. Please remove the namespace.

File: cpp/modmesh/io/gmsh.hpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+{
+struct ElementDef
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: What about naming it like `GmshElementDef`?

File: cpp/modmesh/io/gmsh.hpp [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    small_vector<uint8_t> mmcl() const { return m_mmcl; }
+
+private:
+    uint8_t m_ndim = 0; /* Number of dimension */
+    uint16_t m_nnds = 0; /* Number of nodes     */
+    uint8_t m_mmtpn = 0; /* modmesh cell type   */
+    small_vector<uint8_t> m_mmcl; /* modmesh cell order  */
+}; /* end struct ElementDef */
+
+inline ElementDef ElementDef::by_id(uint16_t id)
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: This is not a template.  Please move it to a cpp file.

File: cpp/modmesh/io/gmsh.hpp [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+{
+#define VEC(...) __VA_ARGS__
+#define MM_DECL_SWITCH_ELEMENT_TYPE(ID, NDIM, NNDS, MMTPN, MMCL) \
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: The use of macro makes the code easy to read.  Thanks.

File: cpp/modmesh/io/gmsh.hpp [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    using number_base = NumberBase<int32_t, double>;
+    using int_type = typename number_base::int_type;
+    using uint_type = typename number_base::uint_type;
+    using real_type = typename number_base::real_type;
+
+public:
+    explicit Gmsh(const std::string & file_path)
+    {
+        bool meta_enter = false, node_enter = false, element_enter = false;
+        // clang-format off
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: You need to turn clang-format back on!!

File: cpp/modmesh/io/gmsh.hpp [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+    {
+        BEGIN,
+        META_END,
+        PYHSICAL_NAME_END,
+        NODE_END,
+        ELEMENT_END
+    };
+
+    // Check the finite state machine state transition is valid or not to check msh file format is correct
+    bool _isValidTransition(const std::string s)
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: Do not prepend `_` to any C++ symbols.  You already made them private and mixing the C++ access with Python private convention is a bad idea.

File: cpp/modmesh/io/gmsh.hpp [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+class Gmsh
+    : public NumberBase<int32_t, double>
+{
+    using number_base = NumberBase<int32_t, double>;
+    using int_type = typename number_base::int_type;
+    using uint_type = typename number_base::uint_type;
+    using real_type = typename number_base::real_type;
+
+public:
+    explicit Gmsh(const std::string & file_path)
```
Comment (yungyuc) [inline cpp/modmesh/io/gmsh.hpp]: Could you make it also take a string as data, in addition to a string as path?  Both are useful.

File: gtests/test_nopython_io.cpp [hunk 12]
```diff
@@ -0,0 +1,350 @@
+#include <modmesh/io/io.hpp>
```
Comment (yungyuc) [inline gtests/test_nopython_io.cpp]: Too long to review.  I trust you have made comprehensive testing.  If not, please speak up.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Thanks for the great work.  Let's polish the code.

File: gtests/test_nopython_io.cpp (hunk 12 above)
Comment (j8xixo12) [reply gtests/test_nopython_io.cpp]: I have tested each cell type conversion between `gmsh` and `modmesh`. Additionally, I also tested if a user inputs a cell type that is not listed in `GmshElementDef`. The expected behavior is that `GmshElement::by_id` should return an empty `GmshElementDef` object.

File: cpp/modmesh/io/CMakeLists.txt (hunk 3 above)
Comment (j8xixo12) [reply cpp/modmesh/io/CMakeLists.txt]: I renamed the module name to `inout` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

File: cpp/modmesh/io/gmsh.hpp (hunk 4 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: I renamed the namespace to `inout` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f too.

File: cpp/modmesh/io/gmsh.hpp (hunk 5 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: I have removed namespace `Gmsh` in  https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

File: cpp/modmesh/io/gmsh.hpp (hunk 6 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: I renamed `ElementDef` to `GmshElementDef` in https://github.com/solvcon/modmesh/pull/233/commits/7432534167ee1da998cfb1a7024ed5e8f52f869f

File: cpp/modmesh/io/gmsh.hpp (hunk 11 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: I change it to use string as data and python can take care of file stream to `bytes`, even if user is using `Gmsh` in cpp, cpp also can take care of file stream to string.

File: cpp/modmesh/io/gmsh.hpp (hunk 7 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: `GmshElementDef::by_id` is a inline function should I move it to cpp file?
This function will be accessed by gtest test case, if move it to cpp file, gtest will encounter undefined reference.

File: cpp/modmesh/io/gmsh.hpp (hunk 10 above)
Comment (j8xixo12) [reply cpp/modmesh/io/gmsh.hpp]: I have removed `_` from these private member function.

File: cpp/modmesh/io/CMakeLists.txt (hunk 3 above)
Comment (yungyuc) [reply cpp/modmesh/io/CMakeLists.txt]: Thanks

File: cpp/modmesh/inout/gmsh.hpp [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    {
+        BEGIN,
+        META_END,
+        PYHSICAL_NAME_END,
+        NODE_END,
+        ELEMENT_END
+    };
+
+    // Check the finite state machine state transition is valid or not to check msh file format is correct
+    bool isValidTransition(const std::string s)
```
Comment (yungyuc) [inline cpp/modmesh/inout/gmsh.hpp]: Ah, I forgot to remind that C++ function names should follow snake_case.  You used snake_case in other functions.  This looks like an overlook.  Could you change it and review all in your PR?

**Review by yungyuc — CHANGES_REQUESTED** [review]
It's quite close to completion.  Only some minor formatting issues remain.

File: cpp/modmesh/inout/gmsh.hpp (hunk 13 above)
Comment (j8xixo12) [reply cpp/modmesh/inout/gmsh.hpp]: I fixed this in https://github.com/solvcon/modmesh/pull/233/commits/f7de14ff4eb817cc2c858ee14dfcaa4ec52f2a49
I changed `isValidTransition` and `toblock` to snake case in `Gmsh` class

File: cpp/modmesh/inout/pymod/wrap_Gmsh.cpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        namespace py = pybind11; // NOLINT(misc-unused-alias-decls)
+
+        (*this)
+            .def(
+                py::init(
+                    [](const py::bytes & data)
+                    { return std::make_shared<inout::Gmsh>(data); }),
+                py::arg("data"))
+            .def("toblock", &wrapped_type::to_block);
```
Comment (yungyuc) [inline cpp/modmesh/inout/pymod/wrap_Gmsh.cpp]: I guess it's still an overlook here.  It makes sense to align the function names between Python and C++.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Yet a minor mismatch.  Should be a simple fix 🙂

**Review by yungyuc — APPROVED** [review]

**Comment by yungyuc** [comment]
Thanks a lot!

**Pull request merged**